packages/vite-plugin-svelte/src/utils/dependencies.ts

```diff
@@ -0,0 +1,156 @@
+import { log } from './log';
+import path from 'path';
+import fs from 'fs';
+import { createRequire } from 'module';
+
+export function findSvelteDependencies(root: string, cwdFallback = true): SvelteDependency[] {
+	log.debug(`findSvelteDependencies: searching svelte dependencies in ${root}`);
+	const pkgFile = path.join(root, 'package.json');
+	if (!fs.existsSync(pkgFile)) {
+		if (cwdFallback) {
+			const cwd = process.cwd();
+			if (root !== cwd) {
+				log.debug(`no package.json found in vite root ${root}`);
+				return findSvelteDependencies(cwd, false);
+			}
+		}
+		log.debug(`no package.json found, search failed`);
+		return [];
+	}
+
+	const pkg = parsePkg(root);
+	if (!pkg) {
+		return [];
+	}
+
+	const deps = [
+		...Object.keys(pkg.dependencies || {}),
+		...Object.keys(pkg.devDependencies || {})
+	].filter((dep) => !excludeFromScan(dep));
+
+	return getSvelteDependencies(deps, root);
+}
+
+function getSvelteDependencies(
+	deps: string[],
+	pkgDir: string,
+	path: string[] = []
+): SvelteDependency[] {
+	const result = [];
+	const require = createRequire(pkgDir);
+	const resolvedDeps = deps.map((dep) => resolveSvelteDependency(dep, require)).filter(Boolean);
+	// @ts-ignore
+	for (const { pkg, dir } of resolvedDeps) {
+		result.push({ name: pkg.name, pkg, dir, path });
+		if (pkg.dependencies) {
+			let dependencyNames = Object.keys(pkg.dependencies);
+			const circular = dependencyNames.filter((name) => path.includes(name));
+			if (circular.length > 0) {
+				log.warn.enabled &&
+					log.warn(
+						`skipping circular svelte dependencies in automated vite optimizeDeps handling`,
+						circular.map((x) => path.concat(x).join('>'))
+					);
+				dependencyNames = dependencyNames.filter((name) => !path.includes(name));
+			}
+			if (path.length === 3) {
+				log.warn.once(`encountered deep svelte dependency tree ${path.join('>')}`);
+			}
+			result.push(...getSvelteDependencies(dependencyNames, dir, path.concat(pkg.name)));
+		}
+	}
+	return result;
+}
+
+function resolveSvelteDependency(
+	dep: string,
+	require: NodeRequire
+): { dir: string; pkg: Pkg } | void {
+	try {
+		const pkgJson = `${dep}/package.json`;
+		const pkg = require(pkgJson);
+		if (!isSvelte(pkg)) {
+			return;
+		}
+		const dir = path.dirname(require.resolve(pkgJson));
+		return { dir, pkg };
+	} catch (e) {
+		log.debug.once(`dependency ${dep} does not export package.json`, e);
```

Comment:
we should give them some idea of how to address it
```suggestion
		log.debug.once(`package.json of ${dep} does not include itself in exports field. Please ask package maintainer to update`, e);
```

Reply:
improved the warning for svelte dependencies instead. I don't think we should put this out here for every dependency that fails. There may be devDependencies that just don't care and the warning would not never go away

Reply:
`rollup-plugin-svelte` warns for all of them, but gathers the list and then just prints a single warning message at the end: https://github.com/sveltejs/rollup-plugin-svelte/pull/119. That might be a good way to do it so that we let users know there's a package that needs to be updated without spamming them with too many messages